### PR TITLE
[RFC] Proposal to merge the Patatrack development into CMSSW 10.5.x [squashed]

### PR DESCRIPTION
**Pull request by fwyzard**
This should be equivalent to #25353 after fixing conflicts and **squashing** on top of CMSSW_10_4_0 .

Currently, not all the developments we would like to integrate have been merged in the development branch: see cms-patatrack#200 for the status of the Patatrack branch with respect to readiness for merging upstream.

### CODE REVIEW

**Comment by fwyzard** [comment]
enable GPU

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by smuzaffar** [comment]
@fwyzard , jenkins job for GPU based testing is failing right now ( https://cmssdt.cern.ch/jenkins/job/ib-any-integration/32576/console ) . I am looking in to it and retry the job (directly via jenkins once fixed). Normal PR testing is running though

**Comment by fwyzard** [comment]
@VinInn please make a PR for Patatrack to fix the build warnings:
```
RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:830:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:834:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:878:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:830:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:834:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:878:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:830:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:834:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:878:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:830:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:834:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:878:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:831:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:835:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:879:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:830:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:834:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:878:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:831:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:835:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:879:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:830:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:834:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:878:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:830:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:834:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:878:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:830:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:834:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:878:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:831:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:835:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
/data/cmsbuild/jenkins/workspace/ib-any-integration/CMSSW_10_5_X_2019-01-14-1100/src/RecoPixelVertexing/PixelTrackFitting/interface/RiemannFit.h:879:0: warning: ignoring #pragma unroll  [-Wunknown-pragmas]
```

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by fwyzard** [comment]
@smuzaffar , do you know why the **cms-bot/GPU** claims that there are  Errors Found in the Matrix Tests Outputs ?
The [logs](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25647/32689/runTheMatrix-results/runall-report-step123-.log) seem fine:
> ```
> 10824.8_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_pixelTrackingOnlyGPU_2018+HARVESTFull_pixelTrackingOnlyGPU_2018 Step0-PASSED Step1-PASSED Step2-PASSED Step3-PASSED  - time date Fri Jan 18 11:30:39 2019-date Fri Jan 18 11:22:02 2019; exit: 0 0 0 0
> 1 1 1 1 tests passed, 0 0 0 0 failed
> ```

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
@cmsbuild, please test

**Comment by fwyzard** [comment]
@smuzaffar I think I have fixed every possible problem reported by the bot.
Can you please check why this
```
10824.8_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_pixelTrackingOnlyGPU_2018+HARVESTFull_pixelTrackingOnlyGPU_2018 Step0-PASSED Step1-PASSED Step2-PASSED Step3-PASSED  - time date Mon Jan 21 15:17:10 2019-date Mon Jan 21 15:08:36 2019; exit: 0 0 0 0
1 1 1 1 tests passed, 0 0 0 0 failed
```
is considered as `Errors Found` ?

**Comment by smuzaffar** [comment]
@fwyzard , I see these[a]  error messages in https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-25647/32734/runTheMatrix-results/matrixTests.log . As we search for word `ERROR ` in the log file that is why bot marks it as failure. By the way, the command run by bot is `runTheMatrix.py --job-reports -j 38 -l 10824.8` , you canm run it locally to check why it is causing these matrix errors.

```
ignoring relval_upgrade from default matrix
processing relval_2017
ERROR in Matrix
overwritting 10824.7 not allowed
ERROR in Matrix
overwritting 11224.7 not allowed
ERROR in Matrix
overwritting 10824.7 not allowed
```

**Comment by fwyzard** [comment]
Thanks, I'm looking at it.

**Comment by fwyzard** [comment]
@smuzaffar I run the same runTheMatrix command, `runTheMatrix.py --job-reports -j 38 -l 10824.8` locally, but I didn't get those ERROR messages.

Any suggestions ?

**Comment by fwyzard** [comment]
@makortel ...

File: HeterogeneousCore/CUDACore/src/GPUCuda.cc [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+                                      if (status == cudaSuccess) {
+                                        locationSetter(HeterogeneousDeviceId(HeterogeneousDevice::kGPUCuda, deviceId));
+                                        LogTrace("GPUCuda") << "  GPU kernel finished (in callback) device " << deviceId << " CUDA stream " << streamId;
+                                        waitingTaskHolder.doneWaiting(nullptr);
+                                      } else {
+                                        // wrap the exception in a try-catch block to let GDB "catch throw" break on it
+                                        try {
+                                          auto error = cudaGetErrorName(status);
+                                          auto message = cudaGetErrorString(status);
+                                          throw cms::Exception("CUDAError") << "Callback of CUDA stream " << streamId << " in device " << deviceId << " error " << error << ": " << message;
```
Comment (Dr15Jones) [inline HeterogeneousCore/CUDACore/src/GPUCuda.cc]: I think this can be replaced with `std::make_exception_ptr`.

**Review by Dr15Jones — CHANGES_REQUESTED** [review]

**Comment by Dr15Jones** [comment]
> @Dr15Jones @makortel what should we do regarding the coding rules violations reported here, regarding the use of `catch(...)` in `HeterogeneousCore/Producer/src/HeterogeneousEDProducer.cc` and `HeterogeneousCore/CUDACore/src/GPUCuda.cc` ?

I think we could suppress it using https://clang.llvm.org/extra/clang-tidy/#id3
@gartung will the clang-tidy suppression also work with our static analyzer changes?

File: HeterogeneousCore/CUDACore/src/GPUCuda.cc (hunk 1 above)
Comment (fwyzard) [reply HeterogeneousCore/CUDACore/src/GPUCuda.cc]: we were using that before,  but I changed to a `throw` in order to be able to `catch throw` it in GDB

**Comment by fwyzard** [comment]
For the moment I have added a whitelist to out python checks, same as for the `FWCore` packages.

File: HeterogeneousCore/CUDACore/src/GPUCuda.cc (hunk 1 above)
Comment (Dr15Jones) [reply HeterogeneousCore/CUDACore/src/GPUCuda.cc]: You could change `catch(..)` to `catch(cms::Exception const&)` instead, since you know the type.

**Comment by fwyzard** [comment]
by the way @kpedro88, do you have any idea why I do not get the same error when testing locally, on top of the latest IB ?

File: HeterogeneousCore/CUDACore/src/GPUCuda.cc (hunk 1 above)
Comment (fwyzard) [reply HeterogeneousCore/CUDACore/src/GPUCuda.cc]: ah, I see what you mean - sure, that makes sense.

**Comment by gartung** [comment]
> @gartung will the clang-tidy suppression also work with our static analyzer changes?
@Dr15Jones not that I know of.

**Comment by fwyzard** [comment]
Will try again...

**Pull request closed without merging**